File: csharp/AerospikeHelper/AerospikeHelper/Tests/CollectionsLargeList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using NUnit.Framework;
using System.IO;

using Aerospike.Client;

namespace Aerospike.Helper.Collections
{
	[TestFixture]
	public class CollectionsLargeList
	{
		public const String HOST = "127.0.0.1";
		public const int PORT = 3000;
		public const int TIMEOUT = 1000;
		public const int EXPIRY = 1800;
		public const String NS = "test";
		public const String SET = "CollectionsLargeList";
		public const String LIST_BIN = "ListBin";

		private AerospikeClient client;
		private Exception caughtException;

		public CollectionsLargeList ()
		{
		}

		[SetUp]
		public virtual void SetUp() {
			try
			{
				Console.WriteLine("Creating AerospikeClient");
				ClientPolicy clientPolicy = new ClientPolicy();
				clientPolicy.timeout = TIMEOUT;
				client = new AerospikeClient(clientPolicy, HOST, PORT);
				client.writePolicyDefault.expiration = EXPIRY;
				client.writePolicyDefault.recordExistsAction = RecordExistsAction.REPLACE;

			} catch (Exception ex)
			{
				caughtException = ex; 
				Console.WriteLine(string.Format("TestFixtureSetUp failed in {0} - {1} {2}", this.GetType(), caughtException.GetType(), caughtException.Message));
				Console.WriteLine (caughtException.StackTrace);
			}
		}

		[TearDown]
		public virtual void TearDown() {
			Console.WriteLine("Closing AerospikeClient");
			client.Close();
			if (caughtException != null)
			{
				Console.WriteLine(string.Format("TestFixtureSetUp failed in {0} - {1} {2}", this.GetType(), caughtException.GetType(), caughtException.Message));
			} 
		}

		[TestCase]
		public void CDTListOperations(){
			Key key = new Key (NS, SET, "CDT-list-test-key");
			Record record = client.Operate(null, key, ListOperation.Clear("integer-list"));
			IList inputList = new List<Value>();
			inputList.Add(Value.Get(55));
			inputList.Add(Value.Get(77));
			for (int x = 0; x < 100; x++) {
				client.Operate(null, key, ListOperation.Insert("integer-list", 0, Value.Get(x)));
			}
			record = client.Operate(null, key, ListOperation.Size("integer-list"));
			//record = client.Operate(null, key, ListOperation.AppendItems("integer-list", inputList));
		}

		/// <summary>
		/// Simple examples of large list functionality.
		/// </summary>
		[TestCase]
		public void RunSimpleExample()
		{
			Key key = new Key(NS, SET, "setkey");
			string binName = LIST_BIN;

			// Delete record if it already exists.
			client.Delete(null, key);

			// Initialize large set operator.
			Aerospike.Helper.Collections.LargeList llist = new Aerospike.Helper.Collections.LargeList(client, null, key, binName);
			string orig1 = "llistValue1";
			string orig2 = "llistValue2";
			string orig3 = "llistValue3";

			// Write values.
			llist.Add(Value.Get(orig1));
			llist.Add(Value.Get(orig2));
			llist.Add(Value.Get(orig3));

			IDictionary map = llist.GetConfig();

			foreach (DictionaryEntry entry in map)
			{
				Console.WriteLine(entry.Key.ToString() + ',' + entry.Value);
			}

			IList rangeList = llist.Range(Value.Get(orig2), Value.Get(orig3));

			Assert.IsNotNull (rangeList);
		
			Assert.AreEqual (rangeList.Count, 2);

			string v2 = (string) rangeList[0];
			string v3 = (string) rangeList[1];

			if (v2.Equals(orig2) && v3.Equals(orig3))
			{
				Console.WriteLine("Range Query matched: v2=" + orig2 + " v3=" + orig3);
			}
			else
			{
				Assert.Fail("Range Content mismatch. Expected (" + orig2 + ":" + orig3 +
					") Received (" + v2 + ":" + v3 + ")");
			}

			// Remove last value.
			llist.Remove(Value.Get(orig3));

			int size = llist.Size();

			if (size != 2)
			{
				throw new Exception("Size mismatch. Expected 2 Received " + size);
			}

			IList listReceived = llist.Find(Value.Get(orig2));
			string expected = orig2;

			if (listReceived == null)
			{
				Console.WriteLine("Data mismatch: Expected " + expected + " Received null");
				Assert.Fail();
			}

			string stringReceived = (string) listReceived[0];

			if (stringReceived != null && stringReceived.Equals(expected))
			{
				Console.WriteLine("Data matched: namespace=" + key.ns + " set=" + key.setName + " key=" + key.userKey +
					" value=" + stringReceived);
			}
			else
			{
				Console.WriteLine("Data mismatch: Expected " + expected + " Received " + stringReceived);
				Assert.Fail();
			}
		}

		/// <summary>
		/// Use distinct sub-bins for row in largelist bin. 
		/// </summary>
		[TestCase]
		public void RunWithDistinctBins()
		{
			Key key = new Key(NS, SET, "accountId");

			// Delete record if it already exists.
			client.Delete(null, key);	

			// Initialize large list operator.
			Aerospike.Helper.Collections.LargeList list = new Aerospike.Helper.Collections.LargeList(client, null, key, "trades");

			list.Size ();

			// Write trades
			Dictionary<string,Value> dict = new Dictionary<string,Value>();

			DateTime timestamp1 = new DateTime(2014, 6, 25, 12, 18, 43);
			dict["key"] = Value.Get(timestamp1.Ticks);
			dict["ticker"] = Value.Get("IBM");
			dict["qty"] = Value.Get(100);
			dict["price"] = Value.Get(BitConverter.GetBytes(181.82));
			list.Add(Value.Get(dict));

			DateTime timestamp2 = new DateTime(2014, 6, 26, 9, 33, 17);
			dict["key"] = Value.Get(timestamp2.Ticks);
			dict["ticker"] = Value.Get("GE");
			dict["qty"] = Value.Get(500);
			dict["price"] = Value.Get(BitConverter.GetBytes(26.36));
			list.Add(Value.Get(dict));

			DateTime timestamp3 = new DateTime(2014, 6, 27, 14, 40, 19);
			dict["key"] = Value.Get(timestamp3.Ticks);
			dict["ticker"] = Value.Get("AAPL");
			dict["qty"] = Value.Get(75);
			dict["price"] = Value.Get(BitConverter.GetBytes(91.85));
			list.Add(Value.Get(dict));

			// Verify list size
			int size = list.Size();

			Assert.AreEqual (size, 3);

			// Filter on range of timestamps
			DateTime begin = new DateTime(2014, 6, 26);
			DateTime end = new DateTime(2014, 6, 28);
			IList results = list.Range(Value.Get(begin.Ticks), Value.Get(end.Ticks));

			Assert.AreEqual (results.Count, 2);

			// Verify data.
			ValidateWithDistinctBins(results, 0, timestamp2, "GE", 500, 26.36);
			ValidateWithDistinctBins(results, 1, timestamp3, "AAPL", 75, 91.85);

			Console.WriteLine("Data matched.");

			Console.WriteLine("Run large list scan.");
			IList rows = list.Scan();
			foreach (IDictionary row in rows)
			{
				foreach (DictionaryEntry entry in row)
				{
					//Console.WriteLine(entry.Key.ToString());
					//Console.WriteLine(entry.Value.ToString());
				}
			}
			Console.WriteLine("Large list scan complete.");
		}


		public void ValidateWithDistinctBins(IList list, int index, DateTime expectedTime, string expectedTicker, int expectedQty, double expectedPrice)
		{
			IDictionary dict = (IDictionary)list[index];
			DateTime receivedTime = new DateTime((long)dict["key"]);

			Assert.AreEqual (expectedTime, receivedTime);

			string receivedTicker = (string)dict["ticker"];

			Assert.AreEqual (expectedTicker, receivedTicker);

			long receivedQty = (long)dict["qty"];

			Assert.AreEqual (expectedQty, receivedQty);

			double receivedPrice = BitConverter.ToDouble((byte[])dict["price"], 0);

			Assert.AreEqual (expectedPrice, receivedPrice);
		}

		/// <summary>
		/// Use serialized bin for row in largelist bin. 
		/// </summary>
		[TestCase]
		public  void RunWithSerializedBin()
		{
			Key key = new Key(NS, SET, "accountId");

			// Delete record if it already exists.
			client.Delete(null, key);

			// Initialize large list operator.
			Aerospike.Helper.Collections.LargeList list = new Aerospike.Helper.Collections.LargeList(client, null, key, "trades");

			// Write trades
			Dictionary<string, Value> dict = new Dictionary<string, Value>();
			MemoryStream ms = new MemoryStream(500);

			DateTime timestamp1 = new DateTime(2014, 6, 25, 12, 18, 43);
			dict["key"] = Value.Get(timestamp1.Ticks);
			BinaryWriter writer = new BinaryWriter(ms);
			writer.Write("IBM");  // ticker
			writer.Write(100);    // qty
			writer.Write(181.82); // price
			dict["value"] = Value.Get(ms.ToArray());
			list.Add(Value.Get(dict));

			DateTime timestamp2 = new DateTime(2014, 6, 26, 9, 33, 17);
			dict["key"] = Value.Get(timestamp2.Ticks);
			ms.SetLength(0);
			writer = new BinaryWriter(ms);
			writer.Write("GE");  // ticker
			writer.Write(500);   // qty
			writer.Write(26.36); // price
			dict["value"] = Value.Get(ms.ToArray());
			list.Add(Value.Get(dict));

			DateTime timestamp3 = new DateTime(2014, 6, 27, 14, 40, 19);
			dict["key"] = Value.Get(timestamp3.Ticks);
			ms.SetLength(0);
			writer = new BinaryWriter(ms);
			writer.Write("AAPL");  // ticker
			writer.Write(75);      // qty
			writer.Write(91.85);   // price
			dict["value"] = Value.Get(ms.ToArray());
			list.Add(Value.Get(dict));

			// Verify list size
			int size = list.Size();

			Assert.AreEqual (size, 3);

			// Filter on range of timestamps
			DateTime begin = new DateTime(2014, 6, 26);
			DateTime end = new DateTime(2014, 6, 28);
			IList results = list.Range(Value.Get(begin.Ticks), Value.Get(end.Ticks));

			Assert.AreEqual (results.Count, 2);

			// Verify data.
			ValidateWithSerializedBin(results, 0, timestamp2, "GE", 500, 26.36);
			ValidateWithSerializedBin(results, 1, timestamp3, "AAPL", 75, 91.85);

			Console.WriteLine("Data matched.");
		}


		public  void ValidateWithSerializedBin(IList list, int index, DateTime expectedTime, string expectedTicker, int expectedQty, double expectedPrice)
		{
			IDictionary dict = (IDictionary)list[index];
			DateTime receivedTime = new DateTime((long)dict["key"]);

			Assert.AreEqual (expectedTime, receivedTime);

			byte[] value = (byte[])dict["value"];
			MemoryStream ms = new MemoryStream(value);
			BinaryReader reader = new BinaryReader(ms);
			string receivedTicker = reader.ReadString();

			Assert.AreEqual (expectedTicker, receivedTicker);

			int receivedQty = reader.ReadInt32();

			Assert.AreEqual (expectedQty, receivedQty);

			double receivedPrice = reader.ReadDouble();

			Assert.AreEqual (expectedPrice, receivedPrice);
		}

		/// <summary>
		/// Use default serialized bin for row in largelist bin. 
		/// </summary>
		/// 
		[TestCase]
		public void RunWithDefaultSerializedBin()
		{
			Key key = new Key(NS, SET, "accountId");
			object value1 = new CompoundObject("IBM", 100);
			object value2 = new CompoundObject("GE", 500);
			object value3 = new CompoundObject("AAPL", 75);

			// Delete record if it already exists.
			client.Delete(null, key);

			// Initialize large list operator.
			Aerospike.Helper.Collections.LargeList list = new Aerospike.Helper.Collections.LargeList(client, null, key, "trades");

			// Write trades
			Dictionary<string, Value> dict = new Dictionary<string, Value>();

			DateTime timestamp1 = new DateTime(2014, 6, 25, 12, 18, 43);
			dict["key"] = Value.Get(timestamp1.Ticks);
			dict["value"] = Value.Get(value1);
			list.Add(Value.Get(dict));

			DateTime timestamp2 = new DateTime(2014, 6, 26, 9, 33, 17);
			dict["key"] = Value.Get(timestamp2.Ticks);
			dict["value"] = Value.Get(value2);
			list.Add(Value.Get(dict));

			DateTime timestamp3 = new DateTime(2014, 6, 27, 14, 40, 19);
			dict["key"] = Value.Get(timestamp3.Ticks);
			dict["value"] = Value.Get(value3);
			list.Add(Value.Get(dict));

			// Verify list size
			int size = list.Size();

			Assert.AreEqual (size, 3);

			// Filter on range of timestamps
			DateTime begin = new DateTime(2014, 6, 26);
			DateTime end = new DateTime(2014, 6, 28);
			IList results = list.Range(Value.Get(begin.Ticks), Value.Get(end.Ticks));

			Assert.AreEqual (results.Count, 2); 

			// Verify data.
			ValidateDefault(results, 0, timestamp2, value2);
			ValidateDefault(results, 1, timestamp3, value3);

			Console.WriteLine("Data matched.");
		}

		private static void ValidateDefault(IList list, int index, DateTime expectedTime, object expectedValue)
		{
			IDictionary dict = (IDictionary)list[index];
			DateTime receivedTime = new DateTime((long)dict["key"]);

			Assert.AreEqual (expectedTime, receivedTime);

			object receivedValue = dict["value"];

			Assert.AreNotEqual (receivedValue, expectedValue);
		}

	}

	[Serializable]
	class CompoundObject
	{
		public string a;
		public int b;

		public CompoundObject(string a, int b)
		{
			this.a = a;
			this.b = b;
		}

		public override bool Equals(object other)
		{
			CompoundObject o = (CompoundObject)other;
			return this.a.Equals(o.a) && this.b == o.b;
		}

		public override int GetHashCode()
		{
			return a.GetHashCode() + b;
		}
	}

}

File: java/src/test/java/com/aerospike/helper/collections/CollectionsLargeList.java
package com.aerospike.helper.collections;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.log4j.Logger;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import com.aerospike.client.AerospikeClient;
import com.aerospike.client.AerospikeException;
import com.aerospike.client.Key;
import com.aerospike.client.Record;
import com.aerospike.client.Value;
import com.aerospike.client.cdt.ListOperation;
import com.aerospike.client.policy.ClientPolicy;
import com.aerospike.helper.query.TestQueryEngine;

public class CollectionsLargeList {

	public static final String SET = "CollectionsLargeList";
	public static final String LIST_BIN = "ListBin";

	private AerospikeClient client;

	protected static Logger log = Logger.getLogger(LargeList.class);

	public CollectionsLargeList ()
	{
	}

	@Before
	public void setUp() {
		try
		{
			log.info("Creating AerospikeClient");
			ClientPolicy clientPolicy = new ClientPolicy();
			clientPolicy.timeout = TestQueryEngine.TIME_OUT;
			client = new AerospikeClient(clientPolicy, TestQueryEngine.HOST, TestQueryEngine.PORT);
			client.writePolicyDefault.expiration = 1800;
			//client.writePolicyDefault.recordExistsAction = RecordExistsAction.REPLACE;

			Key key = new Key (TestQueryEngine.NAMESPACE, SET, "CDT-list-test-key");
			client.delete(null, key);
			key = new Key(TestQueryEngine.NAMESPACE, SET, "setkey");
			client.delete(null, key);
			key = new Key(TestQueryEngine.NAMESPACE, SET, "accountId");
			client.delete(null, key);

		} catch (Exception ex)
		{
			ex.printStackTrace();
		}
	}

	@After
	public void tearDown() {
		log.info("Closing AerospikeClient");
		client.close();
	}


	private void writeIntSubElements(com.aerospike.helper.collections.LargeList ll, int number){
		for (int x = 0; x < number; x++) {
			ll.add(Value.get(x));
		}
		Assert.assertEquals (number, ll.size ());
	}
	private void writeStringSubElements(com.aerospike.helper.collections.LargeList ll, int number){
		for (int x = 0; x < number; x++) {
			ll.add(Value.get("cats-dogs-"+x));
		}
		Assert.assertEquals (number, ll.size ());
	}

	@Test
	public void add100IntOneByOne(){
		Key key = new Key (TestQueryEngine.NAMESPACE, SET, "100-list-test-key-int");
		client.delete(null, key);
		com.aerospike.helper.collections.LargeList ll = new com.aerospike.helper.collections.LargeList (client, null, key, "100-int");
		writeIntSubElements (ll, 100);
		Assert.assertEquals (100, ll.size ());
		ll.destroy ();
		client.delete(null, key);
	}

	@Test
	public void add100StringOneByOne(){
		Key key = new Key (TestQueryEngine.NAMESPACE, SET, "100-list-test-key-String");
		client.delete(null, key);
		com.aerospike.helper.collections.LargeList ll = new com.aerospike.helper.collections.LargeList (client, null, key, "100-String");
		writeStringSubElements (ll, 100);
		Assert.assertEquals (100, ll.size ());
		ll.destroy ();
		client.delete(null, key);


	}

	@Test
	public void scanInt(){
		Key key = new Key (TestQueryEngine.NAMESPACE, SET, "100-list-test-key-int");
		client.delete(null, key);
		com.aerospike.helper.collections.LargeList ll = new com.aerospike.helper.collections.LargeList (client, null, key, "100-int");
		writeIntSubElements (ll, 100);
		List<?> values = ll.scan ();
		Assert.assertEquals (100, ll.size ());
		for (int x = 0; x < 100; x++) {
			Assert.assertEquals (values.get(x), (long)x);
		}
		ll.destroy ();
		client.delete(null, key);

	}
	@Test
	public void scanString(){
		Key key = new Key (TestQueryEngine.NAMESPACE, SET, "100-list-test-key-String");
		client.delete(null, key);
		com.aerospike.helper.collections.LargeList ll = new com.aerospike.helper.collections.LargeList (client, null, key, "100-String");
		writeStringSubElements (ll, 100);
		Assert.assertEquals (100, ll.size ());
		List<?>values = ll.scan ();
		for (int x = 0; x < 100; x++) {
			Assert.assertEquals (values.get(x), "cats-dogs-"+x);
		}
		ll.destroy ();
		client.delete(null, key);

	}
	@Test
	public void CDTListOperations(){
		Key key = new Key (TestQueryEngine.NAMESPACE, SET, "CDT-list-test-key");
		List<Value> inputList = new ArrayList<Value>();
		inputList.add(Value.get(55));
		inputList.add(Value.get(77));
		for (int x = 0; x < 100; x++) {
			client.operate(null, key, ListOperation.insert("integer-list", 0, Value.get(x)));
		}
		Record record = client.operate(null, key, ListOperation.size("integer-list"));
		Assert.assertEquals (100, record.getInt("integer-list"));
		record = client.operate(null, key, ListOperation.appendItems("integer-list", inputList));
		record = client.operate(null, key, ListOperation.size("integer-list"));
		Assert.assertEquals (102, record.getInt("integer-list"));

	}

	/**
	 * Simple examples of large list functionality.
	 */
	@Test
	public void runSimpleExample() throws Exception
	{
		Key key = new Key(TestQueryEngine.NAMESPACE, SET, "setkey");
		String binName = LIST_BIN;

		// Delete record if it already exists.
		client.delete(null, key);

		// Initialize large set operator.
		com.aerospike.helper.collections.LargeList llist = new com.aerospike.helper.collections.LargeList(client, null, key, binName);
		String orig1 = "llistValue1";
		String orig2 = "llistValue2";
		String orig3 = "llistValue3";

		// Write values.
		llist.add(Value.get(orig1));
		llist.add(Value.get(orig2));
		llist.add(Value.get(orig3));

		Assert.assertEquals(llist.size(), 3);


		List<?> rangeList = llist.range(Value.get(orig2), Value.get(orig3));

		Assert.assertNotNull(rangeList);

		Assert.assertEquals (rangeList.size(), 2);

		String v2 = (String) rangeList.get(0);
		String v3 = (String) rangeList.get(1);

		if (v2.equals(orig2) && v3.equals(orig3))
		{
			log.info("Range Query matched: v2=" + orig2 + " v3=" + orig3);
		}
		else
		{
			Assert.fail("Range Content mismatch. Expected (" + orig2 + ":" + orig3 +
					") Received (" + v2 + ":" + v3 + ")");
		}

		// Remove last value.
		llist.remove(Value.get(orig3));

		int size = llist.size();

		if (size != 2)
		{
			throw new Exception("Size mismatch. Expected 2 Received " + size);
		}

		List<?> listReceived = llist.find(Value.get(orig2));
		String expected = orig2;

		if (listReceived == null)
		{
			log.info("Data mismatch: Expected " + expected + " Received null");
			Assert.fail();
		}

		String StringReceived = (String) listReceived.get(0);

		if (StringReceived != null && StringReceived.equals(expected))
		{
			log.info("Data matched: namespace=" + key.namespace + " set=" + key.setName + " key=" + key.userKey +
					" value=" + StringReceived);
		}
		else
		{
			log.info("Data mismatch: Expected " + expected + " Received " + StringReceived);
			Assert.fail();
		}
	}

	/**
	 * Use distinct sub-bins for row in largelist bin. 
	 */
	@Test
	public void runWithDistinctBins()
	{
		Key key = new Key(TestQueryEngine.NAMESPACE, SET, "accountId");

		// Delete record if it already exists.
		client.delete(null, key);	

		// Initialize large list operator.
		com.aerospike.helper.collections.LargeList list = new com.aerospike.helper.collections.LargeList(client, null, key, "trades");

		list.size ();

		// Write trades
		Map<String,Value> map = new HashMap<String,Value>();

		Calendar timestamp1 = new GregorianCalendar(2014, 6, 25, 12, 18, 43);	
		map.put("key", Value.get(timestamp1.getTimeInMillis()));
		map.put("ticker", Value.get("IBM"));
		map.put("qty", Value.get(100));
		map.put("price", Value.get(Double.doubleToLongBits(181.82)));
		list.add(Value.get(map));

		Calendar timestamp2 = new GregorianCalendar(2014, 6, 26, 9, 33, 17);
		map.put("key", Value.get(timestamp2.getTimeInMillis()));
		map.put("ticker", Value.get("GE"));
		map.put("qty", Value.get(500));
		map.put("price", Value.get(Double.doubleToLongBits(26.36)));
		list.add(Value.get(map));

		Calendar timestamp3 = new GregorianCalendar(2014, 6, 27, 14, 40, 19);
		map.put("key", Value.get(timestamp3.getTimeInMillis()));
		map.put("ticker", Value.get("AAPL"));
		map.put("qty", Value.get(75));
		map.put("price", Value.get(Double.doubleToLongBits(91.85)));
		list.add(Value.get(map));

		// Verify list size
		int size = list.size();

		Assert.assertEquals (size, 3);

		// Filter on range of timestamps
		Calendar begin = new GregorianCalendar(2014, 6, 26);
		Calendar end = new GregorianCalendar(2014, 6, 28);
		List<Map<String,Object>> results = (List<Map<String,Object>>)list.range(Value.get(begin.getTimeInMillis()), Value.get(end.getTimeInMillis()));

		Assert.assertEquals (results.size(), 2);

		// Verify data.
		validateWithDistinctBins(results, 0, timestamp2, "GE", 500, 26.36);
		validateWithDistinctBins(results, 1, timestamp3, "AAPL", 75, 91.85);

		log.info("Data matched.");

		log.info("Run large list scan.");
		List<Map<String,Object>> rows = (List<Map<String,Object>>)list.scan();
		for (Map<String,Object> row : rows) {
			for (@SuppressWarnings("unused") Map.Entry<String,Object> entry : row.entrySet()) {
				//console.Info(entry.Key.ToString());
				//console.Info(entry.Value.ToString());
			}
		}
		log.info("Large list scan complete.");
	}


	private void validateWithDistinctBins(List<Map<String,Object>> list, int index, Calendar expectedTime, String expectedTicker, int expectedQty, double expectedPrice) 
	{
		Map<String,Object> map = list.get(index);
		Calendar receivedTime = new GregorianCalendar();
		receivedTime.setTimeInMillis((Long)map.get("key"));

		Assert.assertEquals(expectedTime, receivedTime);

		String receivedTicker = (String)map.get("ticker");

		Assert.assertEquals(expectedTicker, receivedTicker);

		long receivedQty = (Long)map.get("qty");

		Assert.assertEquals(expectedQty, receivedQty) ;

		double receivedPrice = Double.longBitsToDouble((Long)map.get("price"));


		Assert.assertEquals(expectedPrice, receivedPrice, 0.001);
	}

	/**
	 * Use serialized bin for row in largelist bin. 
	 */
	@Test
	public  void runWithSerializedBin() throws AerospikeException, IOException 
	{
		Key key = new Key(TestQueryEngine.NAMESPACE, SET, "accountId");

		// Delete record if it already exists.
		client.delete(null, key);

		// Initialize large list operator.
		com.aerospike.helper.collections.LargeList list = new com.aerospike.helper.collections.LargeList(client, null, key, "trades");

		// Write trades
		Map<String, Value> map = new HashMap<String, Value>();
		ByteArrayOutputStream byteStream = new ByteArrayOutputStream(500);
		DataOutputStream writer = new DataOutputStream(byteStream);

		Calendar timestamp1 = new GregorianCalendar(2014, 6, 25, 12, 18, 43);
		map.put("key", Value.get(timestamp1.getTimeInMillis()));
		writer.writeUTF("IBM");     // ticker
		writer.writeInt(100);       // qty
		writer.writeDouble(181.82); // price
		map.put("value", Value.get(byteStream.toByteArray()));
		list.add(Value.get(map));

		Calendar timestamp2 = new GregorianCalendar(2014, 6, 26, 9, 33, 17);
		map.put("key", Value.get(timestamp2.getTimeInMillis()));
		byteStream.reset();
		writer.writeUTF("GE");     // ticker
		writer.writeInt(500);      // qty
		writer.writeDouble(26.36); // price
		map.put("value", Value.get(byteStream.toByteArray()));
		list.add(Value.get(map));

		Calendar timestamp3 = new GregorianCalendar(2014, 6, 27, 14, 40, 19);
		map.put("key", Value.get(timestamp3.getTimeInMillis()));
		byteStream.reset();
		writer.writeUTF("AAPL");   // ticker
		writer.writeInt(75);       // qty
		writer.writeDouble(91.85); // price
		map.put("value", Value.get(byteStream.toByteArray()));
		list.add(Value.get(map));

		// Verify list size
		int size = list.size();

		Assert.assertEquals (size, 3);

		// Filter on range of timestamps
		Calendar begin = new GregorianCalendar(2014, 6, 26);
		Calendar end = new GregorianCalendar(2014, 6, 28);
		List<Map<String,Object>> results = (List<Map<String,Object>>)list.range(Value.get(begin.getTimeInMillis()), Value.get(end.getTimeInMillis()));

		Assert.assertEquals (results.size(), 2, 0.001);

		// Verify data.
		validateWithSerializedBin(results, 0, timestamp2, "GE", 500, 26.36);
		validateWithSerializedBin(results, 1, timestamp3, "AAPL", 75, 91.85);

		log.info("Data matched.");
	}


	private void validateWithSerializedBin(List<Map<String,Object>> list, int index, Calendar expectedTime, String expectedTicker, int expectedQty, double expectedPrice)
			throws AerospikeException, IOException {
		Map<String,Object> map = list.get(index);
		Calendar receivedTime = new GregorianCalendar();
		receivedTime.setTimeInMillis((Long)map.get("key"));

		Assert.assertEquals (expectedTime, receivedTime);

		byte[] value = (byte[])map.get("value");
		ByteArrayInputStream ms = new ByteArrayInputStream(value);
		DataInputStream reader = new DataInputStream(ms);
		String receivedTicker = reader.readUTF();

		Assert.assertEquals (expectedTicker, receivedTicker);

		int receivedQty = reader.readInt();

		Assert.assertEquals (expectedQty, receivedQty);

		double receivedPrice = reader.readDouble();

		Assert.assertEquals (expectedPrice, receivedPrice, 001);
	}


}

